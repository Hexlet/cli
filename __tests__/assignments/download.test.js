// @ts-check

const os = require('os');
const path = require('path');
const fsp = require('fs/promises');
const fse = require('fs-extra');
const nock = require('nock');

const assignmentDownloadCmd = require('../../src/commands/assignment/download.js');
const { initSettings } = require('../../src/config.js');
const { readDirP, getFixturePath, getAssignmentConfig } = require('../helpers/index.js');

const buildLessonUrl = (courseSlug, lessonSlug) => (
  `https://ru.hexlet.io/courses/${courseSlug}/lessons/${lessonSlug}/theory_unit`
);
const buildApiUrl = (courseSlug, lessonSlug) => (
  `https://ru.hexlet.io/api/course/${courseSlug}/lesson/${lessonSlug}/assignment/download`
);

const courseSlug = 'java-advanced';
const lessonSlug = 'multithreading-java';
const commandParts = ['assignment', 'download'];
const lessonUrl = buildLessonUrl(courseSlug, lessonSlug);
const args = {
  lessonUrl, _: commandParts,
};
const lessonArchivePath = getFixturePath(`${lessonSlug}.tar.gz`);

describe('program', () => {
  let hexletConfigPath;
  let customSettings;
  let hexletDir;
  let config;
  let repoName;

  beforeAll(() => {
    nock.disableNetConnect();

    const scope = nock(buildApiUrl(courseSlug, lessonSlug)).persist();
    scope
      .post('')
      .replyWithFile(200, lessonArchivePath);
  });

  beforeEach(async () => {
    const tmpDir = await fsp.mkdtemp(path.join(os.tmpdir(), 'hexlet-cli-'));
    customSettings = { homedir: tmpDir };
    const settings = initSettings(customSettings);
    hexletConfigPath = settings.hexletConfigPath;
    repoName = settings.repo.name;
    await fse.ensureDir(settings.hexletConfigDir);
    hexletDir = path.join(tmpDir, 'learning', 'Hexlet');
    config = getAssignmentConfig({ hexletDir });
  });

  it('download', async () => {
    await fse.writeJson(hexletConfigPath, config);

    await assignmentDownloadCmd.handler(args, customSettings);
    expect(await readDirP(hexletDir)).toMatchSnapshot();
  });

  it('download the same assignment again (with backup)', async () => {
    await fse.writeJson(hexletConfigPath, config);
    const coursePath = path.join(hexletDir, repoName, courseSlug);
    const assignmentPath = path.join(coursePath, lessonSlug);
    const someFilePath = path.join(assignmentPath, 'SomeFile.java');
    await fse.outputFile(someFilePath, 'content');

    expect(await readDirP(assignmentPath)).toMatchSnapshot();
    await assignmentDownloadCmd.handler(args, customSettings);
    expect(await readDirP(assignmentPath)).toMatchSnapshot();

    const assignments = await fsp.readdir(coursePath);
    const [assignment, assignmentBackup] = assignments.sort();
    expect(assignments.length).toBe(2);
    expect(assignment).toEqual('multithreading-java');
    expect(assignmentBackup).toMatch(/^multithreading-java_\d{2}\.\d{2}\.\d{2}_\d{2}\.\d{2}\.\d{2}$/);

    const assignmentBackupPath = path.join(coursePath, assignmentBackup);
    expect(await readDirP(assignmentBackupPath)).toMatchSnapshot();
  });

  it('download (without init)', async () => {
    await expect(assignmentDownloadCmd.handler(args, customSettings))
      .rejects.toThrow('no such file or directory');
  });

  it('download with invalid .config.json', async () => {
    await fse.writeJson(hexletConfigPath, {});

    await expect(assignmentDownloadCmd.handler(args, customSettings))
      .rejects.toThrow(`Validation error "${hexletConfigPath}"`);
  });

  it('download with incorrect lessonUrl', async () => {
    await fse.writeJson(hexletConfigPath, config);
    const params = {
      ...args,
      lessonUrl: 'https://domain.com',
    };

    await expect(assignmentDownloadCmd.handler(params, customSettings))
      .rejects.toThrow('Incorrect lessonUrl');
  });

  it('download with negative api responses', async () => {
    await fse.writeJson(hexletConfigPath, config);
    const wrongCourseSlug = 'wrong-course';
    const wrongLessonSlug = 'wrong-lesson';
    const params = {
      ...args,
      lessonUrl: buildLessonUrl(wrongCourseSlug, wrongLessonSlug),
    };

    nock(buildApiUrl(wrongCourseSlug, wrongLessonSlug))
      .post('')
      .reply(404);
    await expect(assignmentDownloadCmd.handler(params, customSettings))
      .rejects.toThrow(`Assignment ${wrongCourseSlug}/${wrongLessonSlug} not found. Check the lessonUrl.`);

    let message = 'Invalid token passed.';
    nock(buildApiUrl(wrongCourseSlug, wrongLessonSlug))
      .post('')
      .reply(401, { message });
    await expect(assignmentDownloadCmd.handler(params, customSettings))
      .rejects.toThrow(message);

    message = 'You do not have permission to download assignment.';
    nock(buildApiUrl(wrongCourseSlug, wrongLessonSlug))
      .post('')
      .reply(422, { message });
    await expect(assignmentDownloadCmd.handler(params, customSettings))
      .rejects.toThrow(message);

    // Unhandled errors
    nock(buildApiUrl(wrongCourseSlug, wrongLessonSlug))
      .post('')
      .reply(500);
    await expect(assignmentDownloadCmd.handler(params, customSettings))
      .rejects.toThrow('Request failed with status code 500');
  });
});
