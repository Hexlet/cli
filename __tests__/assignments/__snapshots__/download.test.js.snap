// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download assignment download 1`] = `
Array [
  Object {
    "name": "hexlet-check.yml",
    "path": "hexlet-assignments/.github/workflows/hexlet-check.yml",
    "size": 903n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/README.md",
    "size": 14n,
  },
  Object {
    "name": ".gitignore",
    "path": "hexlet-assignments/java-advanced/multithreading-java/.gitignore",
    "size": 29n,
  },
  Object {
    "name": "Makefile",
    "path": "hexlet-assignments/java-advanced/multithreading-java/Makefile",
    "size": 36n,
  },
  Object {
    "name": "README.html",
    "path": "hexlet-assignments/java-advanced/multithreading-java/README.html",
    "size": 27470n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/java-advanced/multithreading-java/README.md",
    "size": 3583n,
  },
  Object {
    "name": "ShouldBeCleaned.md",
    "path": "hexlet-assignments/java-advanced/multithreading-java/ShouldBeCleaned.md",
    "size": 0n,
  },
  Object {
    "name": "build.gradle",
    "path": "hexlet-assignments/java-advanced/multithreading-java/build.gradle",
    "size": 417n,
  },
  Object {
    "name": "ThreadExample.java",
    "path": "hexlet-assignments/java-advanced/multithreading-java/examples/ThreadExample.java",
    "size": 2687n,
  },
  Object {
    "name": "fileWithMarkers",
    "path": "hexlet-assignments/java-advanced/multithreading-java/fileWithMarkers",
    "size": 32n,
  },
  Object {
    "name": "gradlew",
    "path": "hexlet-assignments/java-advanced/multithreading-java/gradlew",
    "size": 5766n,
  },
  Object {
    "name": "gradlew.bat",
    "path": "hexlet-assignments/java-advanced/multithreading-java/gradlew.bat",
    "size": 2763n,
  },
  Object {
    "name": "presentation.pdf",
    "path": "hexlet-assignments/java-advanced/multithreading-java/presentation.pdf",
    "size": 1110740n,
  },
  Object {
    "name": "App.java",
    "path": "hexlet-assignments/java-advanced/multithreading-java/src/main/java/exercise/App.java",
    "size": 36n,
  },
  Object {
    "name": "AppTest.java",
    "path": "hexlet-assignments/java-advanced/multithreading-java/src/test/java/exercise/AppTest.java",
    "size": 403n,
  },
]
`;

exports[`Download assignment download the same assignment again (with backup) 1`] = `
Array [
  Object {
    "name": "SomeFile.java",
    "path": "SomeFile.java",
    "size": 7n,
  },
]
`;

exports[`Download assignment download the same assignment again (with backup) 2`] = `
Array [
  Object {
    "name": ".gitignore",
    "path": ".gitignore",
    "size": 29n,
  },
  Object {
    "name": "Makefile",
    "path": "Makefile",
    "size": 36n,
  },
  Object {
    "name": "README.html",
    "path": "README.html",
    "size": 27470n,
  },
  Object {
    "name": "README.md",
    "path": "README.md",
    "size": 3583n,
  },
  Object {
    "name": "ShouldBeCleaned.md",
    "path": "ShouldBeCleaned.md",
    "size": 0n,
  },
  Object {
    "name": "build.gradle",
    "path": "build.gradle",
    "size": 417n,
  },
  Object {
    "name": "ThreadExample.java",
    "path": "examples/ThreadExample.java",
    "size": 2687n,
  },
  Object {
    "name": "fileWithMarkers",
    "path": "fileWithMarkers",
    "size": 32n,
  },
  Object {
    "name": "gradlew",
    "path": "gradlew",
    "size": 5766n,
  },
  Object {
    "name": "gradlew.bat",
    "path": "gradlew.bat",
    "size": 2763n,
  },
  Object {
    "name": "presentation.pdf",
    "path": "presentation.pdf",
    "size": 1110740n,
  },
  Object {
    "name": "App.java",
    "path": "src/main/java/exercise/App.java",
    "size": 36n,
  },
  Object {
    "name": "AppTest.java",
    "path": "src/test/java/exercise/AppTest.java",
    "size": 403n,
  },
]
`;

exports[`Download assignment download the same assignment again (with backup) 3`] = `
Array [
  Object {
    "name": "SomeFile.java",
    "path": "SomeFile.java",
    "size": 7n,
  },
]
`;

exports[`Download latest version of the assignment download 1`] = `
Array [
  Object {
    "name": "hexlet-check.yml",
    "path": "hexlet-assignments/.github/workflows/hexlet-check.yml",
    "size": 903n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/README.md",
    "size": 14n,
  },
  Object {
    "name": ".gitignore",
    "path": "hexlet-assignments/java-advanced/multithreading-java/.gitignore",
    "size": 29n,
  },
  Object {
    "name": "Makefile",
    "path": "hexlet-assignments/java-advanced/multithreading-java/Makefile",
    "size": 36n,
  },
  Object {
    "name": "README.html",
    "path": "hexlet-assignments/java-advanced/multithreading-java/README.html",
    "size": 27470n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/java-advanced/multithreading-java/README.md",
    "size": 3583n,
  },
  Object {
    "name": "ShouldBeCleaned.md",
    "path": "hexlet-assignments/java-advanced/multithreading-java/ShouldBeCleaned.md",
    "size": 0n,
  },
  Object {
    "name": "build.gradle",
    "path": "hexlet-assignments/java-advanced/multithreading-java/build.gradle",
    "size": 417n,
  },
  Object {
    "name": "ThreadExample.java",
    "path": "hexlet-assignments/java-advanced/multithreading-java/examples/ThreadExample.java",
    "size": 2687n,
  },
  Object {
    "name": "fileWithMarkers",
    "path": "hexlet-assignments/java-advanced/multithreading-java/fileWithMarkers",
    "size": 32n,
  },
  Object {
    "name": "gradlew",
    "path": "hexlet-assignments/java-advanced/multithreading-java/gradlew",
    "size": 5766n,
  },
  Object {
    "name": "gradlew.bat",
    "path": "hexlet-assignments/java-advanced/multithreading-java/gradlew.bat",
    "size": 2763n,
  },
  Object {
    "name": "presentation.pdf",
    "path": "hexlet-assignments/java-advanced/multithreading-java/presentation.pdf",
    "size": 1110740n,
  },
  Object {
    "name": "App.java",
    "path": "hexlet-assignments/java-advanced/multithreading-java/src/main/java/exercise/App.java",
    "size": 36n,
  },
  Object {
    "name": "AppTest.java",
    "path": "hexlet-assignments/java-advanced/multithreading-java/src/test/java/exercise/AppTest.java",
    "size": 403n,
  },
]
`;

exports[`Download latest version of the assignment download the same assignment again (with backup) 1`] = `
Array [
  Object {
    "name": "SomeFile.java",
    "path": "SomeFile.java",
    "size": 7n,
  },
]
`;

exports[`Download latest version of the assignment download the same assignment again (with backup) 2`] = `
Array [
  Object {
    "name": ".gitignore",
    "path": ".gitignore",
    "size": 29n,
  },
  Object {
    "name": "Makefile",
    "path": "Makefile",
    "size": 36n,
  },
  Object {
    "name": "README.html",
    "path": "README.html",
    "size": 27470n,
  },
  Object {
    "name": "README.md",
    "path": "README.md",
    "size": 3583n,
  },
  Object {
    "name": "ShouldBeCleaned.md",
    "path": "ShouldBeCleaned.md",
    "size": 0n,
  },
  Object {
    "name": "build.gradle",
    "path": "build.gradle",
    "size": 417n,
  },
  Object {
    "name": "ThreadExample.java",
    "path": "examples/ThreadExample.java",
    "size": 2687n,
  },
  Object {
    "name": "fileWithMarkers",
    "path": "fileWithMarkers",
    "size": 32n,
  },
  Object {
    "name": "gradlew",
    "path": "gradlew",
    "size": 5766n,
  },
  Object {
    "name": "gradlew.bat",
    "path": "gradlew.bat",
    "size": 2763n,
  },
  Object {
    "name": "presentation.pdf",
    "path": "presentation.pdf",
    "size": 1110740n,
  },
  Object {
    "name": "App.java",
    "path": "src/main/java/exercise/App.java",
    "size": 36n,
  },
  Object {
    "name": "AppTest.java",
    "path": "src/test/java/exercise/AppTest.java",
    "size": 403n,
  },
]
`;

exports[`Download latest version of the assignment download the same assignment again (with backup) 3`] = `
Array [
  Object {
    "name": "SomeFile.java",
    "path": "SomeFile.java",
    "size": 7n,
  },
]
`;
